src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,542 @@
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            // Get the process information for the specified pid
+            int size = Marshal.SizeOf<proc_taskallinfo>();
+            proc_taskallinfo info = default(proc_taskallinfo);
+            int result = proc_pidinfo(pid, PROC_PIDTASKALLINFO, 0, &info, size);
+            return (result == size ? new proc_taskallinfo?(info) : null);
+        }
+
+        /// <summary>
+        /// Gets the thread information for the given thread
+        /// </summary>
+        /// <param name="thread">The ID of the thread to query for information</param>
+        /// <returns>
+        /// Returns a valid proc_threadinfo struct for valid threads that the caller
+        /// has permissions to access; otherwise, returns null
+        /// </returns>
+        internal static unsafe proc_threadinfo? GetThreadInfoById(int pid, ulong thread)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            // Negative TIDs are invalid
+            if (thread < 0)
+            {
+                throw new ArgumentOutOfRangeException("thread");
+            }
+
+            // Get the thread information for the specified thread in the specified process
+            int size = Marshal.SizeOf<proc_threadinfo>();
+            proc_threadinfo info = default(proc_threadinfo);
+            int result = proc_pidinfo(pid, PROC_PIDTHREADINFO, (ulong)thread, &info, size);
+            return (result == size ? new proc_threadinfo?(info) : null);
+        }
+
+        internal static unsafe List<KeyValuePair<ulong, proc_threadinfo?>> GetAllThreadsInProcess(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            int result = 0;
+            int size = 20; // start assuming 20 threads is enough
+            ulong[] threadIds = null;
+
+            // We have no way of knowning how many threads the process has (and therefore how big our buffer should be)
+            // so while the return value of the function is the same as our buffer size (meaning it completely filled
+            // our buffer), double our buffer size and try again. This ensures that we don't miss any threads
+            do
+            {
+                threadIds = new ulong[size];
+                fixed (ulong* pBuffer = threadIds)
+                {
+                    result = proc_pidinfo(pid, PROC_PIDLISTTHREADS, 0, pBuffer, Marshal.SizeOf<ulong>() * threadIds.Length);
+                }
+
+                if (result <= 0)
+                {
+                    throw new System.ComponentModel.Win32Exception(Marshal.GetLastWin32Error());
+                }
+                else
+                {
+                    size *= 2;
```

Comment:
Nit: Consider making this checked.  We should never get to the point where it matters, but better safe than sorry.